test/apps/iodemo/io_demo.cc

```diff
@@ -1386,30 +1444,19 @@ class DemoClient : public P2pDemoCommon {
     virtual void dispatch_connection_error(UcxConnection *conn) {
         size_t server_index = get_server_index(conn);
         if (server_index < _server_info.size()) {
-            close_server(server_index, ucs_status_string(conn->ucx_status()));
+            disconnect_server(server_index,
+                              ucs_status_string(conn->ucx_status()));
         }
     }
 
-    void close_server(size_t server_index, const char *reason) {
+    void disconnect_server(size_t server_index, const char *reason) {
         server_info_t& server_info = _server_info[server_index];
-        std::map<const UcxConnection*, size_t>::key_type key = server_info.conn;
 
-        LOG << "terminate connection " << server_info.conn << " due to "
+        LOG << "disconnecting connection " << server_info.conn << " due to "
             << reason;
 
         // Destroying the connection will complete its outstanding operations
-        delete server_info.conn;
-
-        // Don't wait for any more completions on this connection
-        _num_sent -= get_num_uncompleted(server_info);
-
-        // Remove connection pointer
-        _server_index_lookup.erase(key);
-
-        // Remove active servers entry
-        active_servers_remove(server_info.active_index);
-
-        reset_server_info(server_info);
+        new Disconnector(*this, *server_info.conn);
```

Comment:
to align with send/recv operations should be something like
`server_info.conn->disconnect(new ClientDisconenctCallback(this))`

Reply:
then `ClientDisconenctCallback` will have uninitialized/NULL `_conn` field in ctor and then disconnect has to call a setter, IMO this is not so good

Reply:
then add conn param:
`server_info.conn->disconnect(new ClientDisconenctCallback(this, conn))`

Reply:
then caller connection and param can be different, that looks buggy prone